server/src/com/cloud/vm/UserVmManagerImpl.java

```diff
@@ -1158,7 +1158,9 @@ public UserVm addNicToVirtualMachine(AddNicToVMCmd cmd) throws InvalidParameterV
             }
         }
 
-        if(_nicDao.findByNetworkIdAndMacAddress(networkId, macAddress) != null) {
+        macAddress = lookForNextAvailableMacAddressInNetworkIfMacIsInvalid(macAddress, network.getId());
```

Comment:
This may enforce and end up changing any passed mac address. Instead, please do a isBlank check on the macaddress, we should only override that when the macaddress is not provided as part of `addNicToVirtualMachine` API.

Reply:
@rhtyd the method only changes the mac address if it is not valid. If it is provided and not changed it returns the same mac. please see lookForNextAvailableMacAddressInNetworkIfMacIsInvalid() below